### ISSUE TITLE
Issue with optimize_pulse_unitary for large sparse matrices

### ISSUE THREAD

**Issue by michelinesoley** [qutip/qutip#1617]
I am using GRAPE to identify optimal pulse sequences to produce a desired unitary, and am running into an error when I use optimize_pulse_unitary for sparse matrices of size 1296x1296 and larger.  

For example, the following code works for N=2 but fails for N=>3:

```python
import numpy as np
from qutip import *
from scipy.sparse import csr_matrix
import qutip.logging_utils as logging
logger = logging.get_logger()
#Set this to None or logging.WARN for 'quiet' execution
log_level = logging.INFO
#QuTiP control modules
import qutip.control.pulseoptim as cpo
import math

# Set parameters
chi  = 2.2 * 2 * np.pi  # coupling strength
N = 3             # number of cavity states
n_ts = 21 # number of discrete time steps
evo_time = 4.      # Length of time
fid_err_targ = 1e-10 # Target fidelity error
max_iter = 0 # # Maximum iterations for the optimization algorithm
max_wall_time = 0 # Maximum (elapsed) time
min_grad = 1e-20 # Minimum gradient
p_type = 'SINE' # pulse type (RND|ZERO|LIN|SINE|SQUARE|SAW|TRIANGLE)
f_ext = "pulsedata.txt" # output file

# Set operators and Hamiltonians

# operators
a  = tensor(destroy(N), qeye(N), qeye(N), qeye(N), qeye(2), qeye(2), qeye(2), qeye(2))
b = tensor(qeye(N), destroy(N), qeye(N), qeye(N), qeye(2), qeye(2), qeye(2), qeye(2))
c  = tensor(qeye(N), qeye(N), destroy(N), qeye(N), qeye(2), qeye(2), qeye(2), qeye(2))
d = tensor(qeye(N), qeye(N), qeye(N),  destroy(N), qeye(2), qeye(2), qeye(2), qeye(2))
sm = tensor(qeye(N), qeye(N), qeye(N), qeye(N), destroy(2), qeye(2), qeye(2), qeye(2))
zm = tensor(qeye(N), qeye(N), qeye(N), qeye(N), qeye(2), destroy(2), qeye(2), qeye(2))
tm = tensor(qeye(N), qeye(N), qeye(N), qeye(N), qeye(2), qeye(2), destroy(2), qeye(2))
qm = tensor(qeye(N), qeye(N), qeye(N), qeye(N), qeye(2), qeye(2), qeye(2), destroy(2))
P_se = tensor(qeye(N), qeye(N), qeye(N), qeye(N), basis(2, 1) * basis(2, 1).dag(), qeye(2), qeye(2), qeye(2))
P_ze = tensor(qeye(N), qeye(N), qeye(N), qeye(N), qeye(2), basis(2, 1) * basis(2, 1).dag(), qeye(2), qeye(2))
P_te = tensor(qeye(N), qeye(N), qeye(N), qeye(N), qeye(2), qeye(2), basis(2, 1) * basis(2, 1).dag(), qeye(2))
P_qe = tensor(qeye(N), qeye(N), qeye(N), qeye(N), qeye(2), qeye(2), qeye(2), basis(2, 1) * basis(2, 1).dag())

# Hamiltonians
drift = chi * (a.dag() * a * P_se) + chi * (b.dag() * b * P_ze) + chi * (c.dag() * c * P_te) + chi * (d.dag() * d * P_qe)

row = np.array([1,1,2,2,3,3,3,3,4,4,4,4,5,5,6,6])
column = np.array([3,4,3,4,1,2,5,6,1,2,5,6,3,4,3,4])
data = np.array([-1,1,-1,-1,-1,-1,-1,-1,1,-1,-1,1,-1,-1,-1,1])

mapped=csr_matrix((data,(row,column)),shape=(N**4*2*2*2*2,N**4*2*2*2*2))
mapped=Qobj(mapped)

n_ctrls=4 # Controls
ctrls = [sm.dag()*a+sm*a.dag()+sm.dag()*b+sm*b.dag()+zm.dag()*b+zm*b.dag()+zm.dag()*c+zm*c.dag()+tm.dag()*c+tm*c.dag()+tm.dag()*d+tm*d.dag()+qm.dag()*d+qm*d.dag()+qm.dag()*a+qm*a.dag(),1j*(sm.dag()*a-sm*a.dag())+1j*(sm.dag()*b-sm*b.dag())+1j*(zm.dag()*b-zm*b.dag())+1j*(zm.dag()*c-zm*c.dag())+1j*(tm.dag()*c-tm*c.dag())+1j*(tm.dag()*d-tm*d.dag())+1j*(qm.dag()*d-qm*d.dag())+1j*(qm.dag()*a-qm*a.dag()),sm+sm.dag()+zm+zm.dag()+tm+tm.dag()+qm+qm.dag(), 1j*(sm-sm.dag())+1j*(zm-zm.dag())+1j*(tm-tm.dag())+1j*(qm-qm.dag())]

# Set optimization parameters
U_0=tensor(qeye(N), qeye(N), qeye(N), qeye(N), qeye(2), qeye(2), qeye(2), qeye(2)) # Initial gate
U_targ=mapped # Target gates

# Run the optimization

result = cpo.optimize_pulse_unitary(drift, ctrls, U_0, U_targ, n_ts, evo_time, fid_err_targ=fid_err_targ, min_grad=min_grad, max_iter=max_iter, max_wall_time=max_wall_time, out_file_ext=f_ext, init_pulse_type=p_type, log_level=log_level,gen_stats=True,dyn_params=dict(sparse_eigen_decomp=True))

```
For N=>3, instead of getting a printout of the optimized controls, I receive the following terminal output error:
```
Traceback (most recent call last):
  File "MinimalErrorExample.py", line 59, in <module>
    result = cpo.optimize_pulse_unitary(drift, ctrls, U_0, U_targ, n_ts, evo_time, fid_err_targ=fid_err_targ, min_grad=min_grad, max_iter=max_iter, max_wall_time=max_wall_time, out_file_ext=f_ext, init_pulse_type=p_type, log_level=log_level,gen_stats=True,dyn_params=dict(sparse_eigen_decomp=True))
  File "/Users/msoley/opt/anaconda3/lib/python3.7/site-packages/qutip/control/pulseoptim.py", line 829, in optimize_pulse_unitary
    gen_stats=gen_stats)
  File "/Users/msoley/opt/anaconda3/lib/python3.7/site-packages/qutip/control/pulseoptim.py", line 510, in optimize_pulse
    result = optim.run_optimization()
  File "/Users/msoley/opt/anaconda3/lib/python3.7/site-packages/qutip/control/optimizer.py", line 970, in run_optimization
    result = self._create_result()
  File "/Users/msoley/opt/anaconda3/lib/python3.7/site-packages/qutip/control/optimizer.py", line 375, in _create_result
    result.initial_fid_err = self.dynamics.fid_computer.get_fid_err()
  File "/Users/msoley/opt/anaconda3/lib/python3.7/site-packages/qutip/control/fidcomp.py", line 448, in get_fid_err
    return np.abs(1 - self.get_fidelity())
  File "/Users/msoley/opt/anaconda3/lib/python3.7/site-packages/qutip/control/fidcomp.py", line 458, in get_fidelity
    self.fid_norm_func(self.get_fidelity_prenorm())
  File "/Users/msoley/opt/anaconda3/lib/python3.7/site-packages/qutip/control/fidcomp.py", line 474, in get_fidelity_prenorm
    dyn.compute_evolution()
  File "/Users/msoley/opt/anaconda3/lib/python3.7/site-packages/qutip/control/dynamics.py", line 1421, in compute_evolution
    self.tslot_computer.recompute_evolution()
  File "/Users/msoley/opt/anaconda3/lib/python3.7/site-packages/qutip/control/tslotcomp.py", line 299, in recompute_evolution
    prop_comp._compute_prop_grad(k, j)
  File "/Users/msoley/opt/anaconda3/lib/python3.7/site-packages/qutip/control/propcomp.py", line 272, in _compute_prop_grad
    dyn._ensure_decomp_curr(k)
  File "/Users/msoley/opt/anaconda3/lib/python3.7/site-packages/qutip/control/dynamics.py", line 1437, in _ensure_decomp_curr
    self._spectral_decomp(k)
  File "/Users/msoley/opt/anaconda3/lib/python3.7/site-packages/qutip/control/dynamics.py", line 1654, in _spectral_decomp
    dims=self.dyn_dims)
  File "/Users/msoley/opt/anaconda3/lib/python3.7/site-packages/qutip/qobj.py", line 292, in __init__
    _tmp = sp.csr_matrix(inpt, dtype=complex, copy=do_copy)
  File "/Users/msoley/opt/anaconda3/lib/python3.7/site-packages/scipy/sparse/compressed.py", line 88, in __init__
    self._set_self(self.__class__(coo_matrix(arg1, dtype=dtype)))
  File "/Users/msoley/opt/anaconda3/lib/python3.7/site-packages/scipy/sparse/coo.py", line 191, in __init__
    self.row, self.col = M.nonzero()
  File "/Users/msoley/opt/anaconda3/lib/python3.7/site-packages/scipy/sparse/base.py", line 287, in __bool__
    raise ValueError("The truth value of an array with more than one "
ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all().

```

I am using the following QuTiP environment:
QuTiP Version:      4.4.1
Numpy Version:      1.21.0
Scipy Version:      1.4.1
Cython Version:     0.29.15
Matplotlib Version: 3.4.2
Python Version:     3.7.6
Number of CPUs:     2
BLAS Info:          OPENBLAS
OPENMP Installed:   False
INTEL MKL Ext:      False
Platform Info:      Darwin (x86_64)

Thank you for your help.

**Comment by hodgestar** [comment]
Hi @michelinesoley,

Your bug report inspired me to dig into the `qutip.control.optimize_pulse` sparse matrix support which lead to some fixes in #1621.

When you increase `N` in your example from 2 to 3, it changes the internal type used by the optimizer to represent propagators. You can also make the `N = 2` case fail by changing the last line to:

```python
result = cpo.optimize_pulse_unitary(..., dyn_params=dict(oper_dtype=Qobj, sparse_eigen_decomp=True))
```

Your example can be made to work for `N = 3` (albeit slowly) by changing the following lines:

```python
mapped = Qobj(mapped,  dims=[[N]*4 + [2]*4] * 2)
```
(which was a bug in the example itself -- it created a Qobj with the right shape but wrong dimensions) and the last line to 

```python
result = cpo.optimize_pulse_unitary(..., dyn_params=dict(oper_dtype=Qobj, sparse_eigen_decomp=False))
```
For me the `N = 3` case was slower with `sparse_eigen_decomp=True`, but perhaps it will be better for your full problem. If you would like to try `sparse_eigen_decomp=True` you are welcome to try the patch in PR #1621 and I would appreciate feedback on any performance issues for larger systems.

We are currently working to support bigger systems in QuTiP 5 by using the new flexible Qobj data layer (which supports both sparse and dense matrices already and will hopefully support GPU and TensorFlow matrices in the near future).

**Comment by michelinesoley** [comment]
Thank you so much all of your help and for responding so quickly!  The fix to the example works very well, thanks!

I am working on getting the patch working and installed commits ecca45c, 3e908bb, 8ce7757, 03b4cb7, 346b2e7, and a1c809a. I am getting a ValueError when I run the example and think I might have accidentally missed a commit - would you have an idea of what I might be missing?  I am including my QuTiP version information below.  Thanks for all of your help.

QuTiP Version:      4.7.0.dev0+a1c809a
Numpy Version:      1.21.0
Scipy Version:      1.4.1
Cython Version:     0.29.24
Matplotlib Version: 3.4.2
Python Version:     3.7.6
Number of CPUs:     2
BLAS Info:          OPENBLAS
OPENMP Installed:   False
INTEL MKL Ext:      False
Platform Info:      Darwin (x86_64)

**Comment by hodgestar** [comment]
@michelinesoley Thank you for trying the PR. Is it possible for you to checkout the exact commit from the branch (i.e. ecca45c8046b49c978224f51662520304aa4e6b7) and try that? If the ValueError persists, could you post the end of traceback here?

**Comment by michelinesoley** [comment]
Thanks!  I used git checkout for ecca45c and reinstalled with python setup.py install, and I get the following error for the minimal example:
Traceback (most recent call last):
  File "MinimalErrorExample.py", line 62, in <module>
    result = cpo.optimize_pulse_unitary(drift, ctrls, U_0, U_targ, n_ts, evo_time, fid_err_targ=fid_err_targ, min_grad=min_grad, max_iter=max_iter, max_wall_time=max_wall_time, out_file_ext=f_ext, init_pulse_type=p_type, log_level=log_level,gen_stats=True,dyn_params=dict(oper_dtype=Qobj,sparse_eigen_decomp=True))
  File "/Users/msoley/opt/anaconda3/lib/python3.7/site-packages/qutip/control/pulseoptim.py", line 757, in optimize_pulse_unitary
    gen_stats=gen_stats)
  File "/Users/msoley/opt/anaconda3/lib/python3.7/site-packages/qutip/control/pulseoptim.py", line 459, in optimize_pulse
    result = optim.run_optimization()
  File "/Users/msoley/opt/anaconda3/lib/python3.7/site-packages/qutip/control/optimizer.py", line 957, in run_optimization
    result = self._create_result()
  File "/Users/msoley/opt/anaconda3/lib/python3.7/site-packages/qutip/control/optimizer.py", line 362, in _create_result
    result.initial_fid_err = self.dynamics.fid_computer.get_fid_err()
  File "/Users/msoley/opt/anaconda3/lib/python3.7/site-packages/qutip/control/fidcomp.py", line 412, in get_fid_err
    return np.abs(1 - self.get_fidelity())
  File "/Users/msoley/opt/anaconda3/lib/python3.7/site-packages/qutip/control/fidcomp.py", line 422, in get_fidelity
    self.fid_norm_func(self.get_fidelity_prenorm())
  File "/Users/msoley/opt/anaconda3/lib/python3.7/site-packages/qutip/control/fidcomp.py", line 438, in get_fidelity_prenorm
    dyn.compute_evolution()
  File "/Users/msoley/opt/anaconda3/lib/python3.7/site-packages/qutip/control/dynamics.py", line 1394, in compute_evolution
    self.tslot_computer.recompute_evolution()
  File "/Users/msoley/opt/anaconda3/lib/python3.7/site-packages/qutip/control/tslotcomp.py", line 265, in recompute_evolution
    prop_comp._compute_prop_grad(k, j)
  File "/Users/msoley/opt/anaconda3/lib/python3.7/site-packages/qutip/control/propcomp.py", line 239, in _compute_prop_grad
    dyn._ensure_decomp_curr(k)
  File "/Users/msoley/opt/anaconda3/lib/python3.7/site-packages/qutip/control/dynamics.py", line 1410, in _ensure_decomp_curr
    self._spectral_decomp(k)
  File "/Users/msoley/opt/anaconda3/lib/python3.7/site-packages/qutip/control/dynamics.py", line 1627, in _spectral_decomp
    dims=self.dyn_dims)
  File "/Users/msoley/opt/anaconda3/lib/python3.7/site-packages/qutip/qobj.py", line 273, in __init__
    _tmp = sp.csr_matrix(inpt, dtype=complex, copy=do_copy)
  File "/Users/msoley/opt/anaconda3/lib/python3.7/site-packages/scipy/sparse/compressed.py", line 88, in __init__
    self._set_self(self.__class__(coo_matrix(arg1, dtype=dtype)))
  File "/Users/msoley/opt/anaconda3/lib/python3.7/site-packages/scipy/sparse/coo.py", line 191, in __init__
    self.row, self.col = M.nonzero()
  File "/Users/msoley/opt/anaconda3/lib/python3.7/site-packages/scipy/sparse/base.py", line 287, in __bool__
    raise ValueError("The truth value of an array with more than one "
ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all().

Is there something I am missing in my installation?  Thanks again for all of your help.

**Comment by hodgestar** [comment]
#1621 has been merged and should be included in the 4.6.3 release.

**Comment by michelinesoley** [comment]
Thank you for all of your help!
